packages/vite-plugin/src/index.ts

```diff
@@ -69,100 +88,105 @@ export function vanillaExtractPlugin({
       // imported from outside the config root.
-      const absoluteId = source.startsWith(config.root)
-        ? source
-        : getAbsoluteVirtualFileId(validId);
+      const absoluteId = getAbsoluteFileId(validId);
 
       // There should always be an entry in the `cssMap` here.
       // The only valid scenario for a missing one is if someone had written
       // a file in their app using the .vanilla.js/.vanilla.css extension
-      if (cssMap.has(absoluteId)) {
+      if (
+        compiler?.getCssForFile(virtualIdToFileId(absoluteId)) ||
+        cssMap.has(absoluteId)
+      ) {
         // Keep the original query string for HMR.
         return absoluteId + (query ? `?${query}` : '');
       }
     },
     load(id) {
       const [validId] = id.split('?');
 
-      if (!cssMap.has(validId)) {
-        return;
-      }
+      if (!validId.endsWith(virtualExtCss)) return;
 
-      const css = cssMap.get(validId);
+      if (compiler) {
+        const absoluteId = getAbsoluteFileId(validId);
 
-      if (typeof css !== 'string') {
-        return;
-      }
+        let { css } =
+          compiler.getCssForFile(virtualIdToFileId(absoluteId)) ?? {};
 
-      if (validId.endsWith(virtualExtCss)) {
         return css;
       }
 
-      return outdent`
-        import { injectStyles } from '@vanilla-extract/css/injectStyles';
-
-        const inject = (css) => injectStyles({
-          fileScope: ${JSON.stringify({ filePath: validId })},
-          css
-        });
-
-        inject(${JSON.stringify(css)});
+      const css = cssMap.get(validId);
 
-        if (import.meta.hot) {
-          import.meta.hot.on('${styleUpdateEvent(validId)}', (css) => {
-            inject(css);
-          });
-        }
-      `;
+      return css;
     },
-    async transform(code, id, ssrParam) {
+    async transform(code, id) {
       const [validId] = id.split('?');
 
       if (!cssFileFilter.test(validId)) {
         return null;
       }
 
-      const identOption =
-        identifiers ?? (config.mode === 'production' ? 'short' : 'debug');
+      if (compiler) {
+        const absoluteId = getAbsoluteFileId(validId);
 
-      let ssr: boolean | undefined;
+        if (DEBUG) console.time(`[compiler] ${validId}`);
+        const { source, watchFiles } = await compiler.processVanillaFile(
+          absoluteId,
+          { outputCss: true },
+        );
+        if (DEBUG) console.timeEnd(`[compiler] ${validId}`);
 
-      if (typeof ssrParam === 'boolean') {
-        ssr = ssrParam;
-      } else {
-        ssr = ssrParam?.ssr;
+        for (const file of watchFiles) {
+          // In start mode, we need to prevent the file from rewatching itself.
+          // If it's a `build --watch`, it needs to watch everything.
+          if (
+            config.command === 'build' ||
+            normalizePath(file) !== absoluteId
+          ) {
+            this.addWatchFile(file);
```

Comment:
Might be worth considering skipping `node_modules`, like the workaround linked in https://github.com/vanilla-extract-css/vanilla-extract/issues/1211:

https://github.com/wenchonglee/ve-mantine-benchmark/blob/main/patches/@vanilla-extract__vite-plugin@3.9.0.patch